### PR TITLE
[feature]: Make import completion status configurable, including per record

### PR DESCRIPTION
**Pull request by deeonwuli**
### :pushpin: References

* **Issue:** Closes [#c6u77a](https://app.clickup.com/t/c6u77a)

### :memo: Implementation
- Completion status on import was previously hardcoded — event/tracker programs always sent `COMPLETED`, data sets never registered completion at all. Added a `markCompletedOnImport` setting plus a per-import checkbox override, covering data sets, event programs, and tracker programs.
- Data set completion works at true registration granularity: one bad data element among hundreds does not block completion of the whole registration (or the whole import).
- Per-record override: a new `Completed` column in the generated templates (new versions `DATASET_GENERATED_v3`, `PROGRAM_GENERATED_v5`, `TRACKER_PROGRAM_GENERATED_v4`) lets a single file carry mixed completion states. A blank cell falls back to the checkbox/setting; old template versions and custom templates are unaffected. The dropdown resolves correctly in any language via the workbook's existing defined-name machinery.
- Download-with-data pre-fills that column with each record's real current completion status — a new `completeDataSetRegistrations` lookup for data sets (opt-in, so the duplicate-check import path doesn't pay for it), and the event/tracker path's already-fetched status — so downloading and re-importing a file unchanged round-trips instead of silently reverting to the default.

### :fire: Notes for the reviewer
 - Download each of the three templates with data for records with mixed real completion states; confirm `Completed` is pre-filled correctly, not blank.
  - Import a file with mixed Yes/No/blank rows; confirm per-record completion in the network tab (`/completeDataSetRegistrations` calls, event `status`).
  - Download with data, re-import unchanged, confirm no completion status flips.
  - Import an old (pre-`Completed`-column) file and a custom template (NHWA/MSF); confirm unchanged behavior.
  - With duplicate-check enabled, confirm `/completeDataSetRegistrations` does **not** fire during import — only during download-with-data.
 
### :video_camera: Screenshots/Screen capture

https://github.com/user-attachments/assets/2e2004b4-6990-43a5-9ffa-91dcb05249d9

https://github.com/user-attachments/assets/2a6b1ec7-2069-45b3-9582-0a3aedc9e65c

Screenshot 2026-07-21 at 17 35 45
Screenshot 2026-07-21 at 17 36 03

#c6u77a

### CODE REVIEW

File: src/domain/entities/DhisDataPackage.ts [hunk 1]
```diff
@@ -1,3 +1,4 @@
+import { EventStatus } from "../../data/Dhis2Import";
```
Comment (anagperal) [inline src/domain/entities/DhisDataPackage.ts]: Domain layer must not depend from data layer. Here you can create a new type in domain.

File: src/utils/booleans.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import { Maybe } from "../types/utils";
+
+const trueValues = ["y", "yes", "true", "1"];
+const falseValues = ["n", "no", "false", "0"];
+
+// Unrecognised input returns undefined rather than false, so callers can fall back to a default.
+export function parseBooleanCell(value: unknown, optionId?: Maybe<string>): Maybe<boolean> {
```
Comment (anagperal) [inline src/utils/booleans.ts:7]: [minor] These two functions are easily testeable

File: src/webapp/logic/sheetBuilder.ts [hunk 3]
```diff
@@ -260,6 +260,15 @@ export class SheetBuilder {
                 } *` + dateFormatInfo
             );
 
+            this.createColumn(
+                sheet,
+                itemRow,
+                columnId++,
+                i18n.t("Completed", { lng: this.builder.language }),
```
Comment (anagperal) [inline src/webapp/logic/sheetBuilder.ts:280]: "Completed" not in translations, run` yarn localize` and commit

File: src/data/InstanceDhisRepository.ts [hunk 4]
```diff
@@ -505,19 +549,84 @@ export class InstanceDhisRepository implements InstanceRepository {
...
+            try {
+                await this.api
+                    .post<{ status: string }>(
+                        "/completeDataSetRegistrations",
+                        {},
+                        { completeDataSetRegistrations: chunk }
+                    )
+                    .getData();
+                return undefined;
+            } catch (error: any) {
```
Comment (anagperal) [inline src/data/InstanceDhisRepository.ts]: [minor] instead of  `any` it's better to use `unknown` + type-guard helper

File: src/domain/usecases/ImportTemplateUseCase.ts [hunk 5]
```diff
@@ -679,6 +682,8 @@ export const compareDataPackages = (
         if (baseValue && compareValue && !areEqual) return false;
     }
 
+    if (base.completed !== undefined) return false;
```
Comment (anagperal) [inline src/domain/usecases/ImportTemplateUseCase.ts]: This bypasses duplicate detection for any row with a completion value. When Duplicate detection is on, it can silently create duplicate events (event/tracker rows without an id that match existing ones). Rather than deciding for the user, we should still surface the DUPLICATE_VALUES warning and let them choose.

I have tested it:
1. Download template from event program with data
2. Delete the event id column and leave Completed column blank:

Image

1. Download template from event program with data
2. Delete the event id column and leave Completed with some value:

Image

Add a test to tthe duplicate detection if you have time to do it

File: src/data/__tests__/Dhis2DataSetCompletion.spec.ts [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+        ...overrides,
+    };
+}
+
+function response(overrides: Partial<DataValueSetsPostResponse> = {}): DataValueSetsPostResponse {
+    return {
+        responseType: "ImportSummary",
+        status: "SUCCESS",
+        description: "",
+        importOptions: {} as DataValueSetsPostResponse["importOptions"],
```
Comment (anagperal) [inline src/data/__tests__/Dhis2DataSetCompletion.spec.ts]: [minor] create variable instead of  assert: 

```
   const defaultImportOptions: DataValueSetsPostResponse["importOptions"] = {
      idSchemes: {},
      dryRun: false,
      async: false,
      importStrategy: "CREATE_AND_UPDATE",
      mergeMode: "REPLACE",
      reportMode: "FULL",
      skipExistingCheck: false,
      sharing: false,
      skipNotifications: false,
      skipAudit: false,
      datasetAllowsPeriods: false,
      strictPeriods: false,
      strictDataElements: false,
      strictCategoryOptionCombos: false,
      strictAttributeOptionCombos: false,
      strictOrganisationUnits: false,
      requireCategoryOptionCombo: false,
      requireAttributeOptionCombo: false,
      skipPatternValidation: false,
      ignoreEmptyCollection: false,
      force: false,
      firstRowIsHeader: false,
      skipLastUpdated: false,
  };
```

File: src/domain/repositories/InstanceRepository.ts [hunk 7]
```diff
@@ -41,13 +42,14 @@ export interface InstanceRepository {
     deleteAggregatedData(dataPackage: DataPackage, rowLookup?: ImportRowLookup): Promise<SynchronizationResult>;
     importDataPackage(dataPackage: DataPackage, options: ImportDataPackageOptions): Promise<SynchronizationResult[]>;
     getProgram(programId: Id): Promise<Program | undefined>;
-    convertDataPackage(dataPackage: DataPackage): EventsPackage | AggregatedPackage;
+    convertDataPackage(dataPackage: DataPackage, markCompleted: boolean): EventsPackage | AggregatedPackage;
```
Comment (anagperal) [inline src/domain/repositories/InstanceRepository.ts:45]: [minor] mismatch: here in the interface markCompleted is required but in the implementation gives a default value

**Review by anagperal — CHANGES_REQUESTED** [review]
thanks @deeonwuli !! Overall this looks solid. it covers a good number of cases (checkbox set, checkbox unset, row present in the Excel, row missing) cleanly. Just a few minor comments below, plus one that does affect functionality.

File: src/domain/usecases/ImportTemplateUseCase.ts (hunk 5 above)
Comment (deeonwuli) [reply src/domain/usecases/ImportTemplateUseCase.ts]: Fixed. I have now removed the `base.completed !== undefined` guard, so completed rows now go through normal duplicate detection. The guard hit data sets too, so the fix covers both types. Thank you!

**Review by anagperal — APPROVED** [review]
thanks @deeonwuli !
